### PR TITLE
Get rid of explicit Git server protocol

### PR DESCRIPTION
**Pull request by fmeum**
## :loudspeaker: Type of change

- [ ] Bugfix
- [ ] New feature
- [x] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## :scroll: Description

~~Ensure that the Git server protocol is not at odds with the URL scheme.~~
    
~~Also move the Protocol switches below the URL to make it clear that the URL should be entered first.~~

~~@msfjarvis We could get rid of the protocol switch alltogether and just extract that information from the URL. I just don't know whether combined with the auth mode choice this wouldn't make the UI more confusing for users.~~
Get rid of the explicit Git server protocol choice which is implicitly available from the repository URL.

## :pencil: Checklist

- [ ] I formatted the code with the IDE's reformat action (Ctrl + Shift + L/Cmd + Shift + L)
- [ ] I reviewed submitted code
- [ ] I added a [CHANGELOG](CHANGELOG.md) entry if applicable

### CODE REVIEW

**Comment by msfjarvis** [comment]
I think we can get rid of the protocol option safely.

**Comment by fmeum** [comment]
> I think we can get rid of the protocol option safely.

How do we ensure that the user only selects an appropriate AuthMode though? Should we just show an error snackbar if the choice is not supported?

**Comment by msfjarvis** [comment]
> > I think we can get rid of the protocol option safely.
> 
> How do we ensure that the user only selects an appropriate AuthMode though? Should we just show an error snackbar if the choice is not supported?

Yep, validate during save. A dialog here would be a better option because the explanation can be a tad wordy and snackbars don't really cut it.

**Comment by fmeum** [comment]
> > > I think we can get rid of the protocol option safely.
> > 
> > 
> > How do we ensure that the user only selects an appropriate AuthMode though? Should we just show an error snackbar if the choice is not supported?
> 
> Yep, validate during save. A dialog here would be a better option because the explanation can be a tad wordy and snackbars don't really cut it.

Sounds reasonable.

I don't think I will have the time to work on this in the near future though as the migration away from JSch and towards ed25519 support turns out to be more work than expected. But I will keep this at the back of my mind.

**Comment by msfjarvis** [comment]
> > > > I think we can get rid of the protocol option safely.
> > > 
> > > 
> > > How do we ensure that the user only selects an appropriate AuthMode though? Should we just show an error snackbar if the choice is not supported?
> > 
> > 
> > Yep, validate during save. A dialog here would be a better option because the explanation can be a tad wordy and snackbars don't really cut it.
> 
> Sounds reasonable.
> 
> I don't think I will have the time to work on this in the near future though as the migration away from JSch and towards ed25519 support turns out to be more work than expected. But I will keep this at the back of my mind.

I'll get this one sorted.

**Comment by msfjarvis** [comment]
@FabianHenneke I've removed the protocol selection and moved the validation to GitSettings. To notify the user, I elected to take a 'shortcut' and let them know of the correct authentication modes for their selected protocol via the respective enums. Since we no longer keep track of protocol in GitServerConfigActivity, I had to convert the update result enum into a sealed class and pass in the protocol and valid authentication modes from GitSettings. This works well in my testing, and I've updated the migrations to drop the protocol field in case someone's doing a big jump from 1.9.x or something.

File: app/src/main/java/com/zeapo/pwdstore/git/config/GitSettings.kt [hunk 1]
```diff
@@ -104,23 +97,40 @@ object GitSettings {
...
-    enum class UpdateConnectionSettingsResult {
-        Valid,
-        FailedToParseUrl,
-        MissingUsername,
+    sealed class UpdateConnectionSettingsResult {
+        class MissingUsername(val newProtocol: Protocol) : UpdateConnectionSettingsResult()
+        class AuthModeMismatch(val newProtocol: Protocol, val validModes: List<AuthMode>) : UpdateConnectionSettingsResult()
+        object Valid : UpdateConnectionSettingsResult()
+        object FailedToParseUrl : UpdateConnectionSettingsResult()
+        object ProtocolMismatch : UpdateConnectionSettingsResult()
```
Comment (fmeum) [inline app/src/main/java/com/zeapo/pwdstore/git/config/GitSettings.kt]: We could get rid of this now, right?
```suggestion
```

File: app/src/main/java/com/zeapo/pwdstore/git/GitServerConfigActivity.kt [hunk 2]
```diff
@@ -78,27 +61,28 @@ class GitServerConfigActivity : BaseGitActivity() {
...
+                is GitSettings.UpdateConnectionSettingsResult.MissingUsername -> {
+                    when (updateResult.newProtocol) {
                         Protocol.Https -> Snackbar.make(binding.root, getString(R.string.git_server_config_save_missing_username_https), Snackbar.LENGTH_LONG).show()
                         Protocol.Ssh -> Snackbar.make(binding.root, getString(R.string.git_server_config_save_missing_username_ssh), Snackbar.LENGTH_LONG).show()
                     }
                 }
-                else -> {
+                GitSettings.UpdateConnectionSettingsResult.ProtocolMismatch -> {
+                    Snackbar.make(binding.root, getString(R.string.git_server_config_save_protocol_mismatch), Snackbar.LENGTH_LONG).show()
+                }
```
Comment (fmeum) [inline app/src/main/java/com/zeapo/pwdstore/git/GitServerConfigActivity.kt]:
```suggestion
```

File: app/src/main/res/values/strings.xml [hunk 3]
```diff
@@ -330,6 +330,8 @@
     <string name="git_server_config_save_error">The provided repository URL is not valid</string>
     <string name="git_server_config_save_missing_username_https">Please specify the HTTPS username in the form https://username@example.com/…</string>
     <string name="git_server_config_save_missing_username_ssh">Please specify the SSH username in the form username@example.com:…</string>
+    <string name="git_server_config_save_protocol_mismatch">The provided repository URL\'s scheme does not match the selected protocol</string>
```
Comment (fmeum) [inline app/src/main/res/values/strings.xml]:
```suggestion
```

**Review by fmeum — COMMENTED** [review]
Thanks for taking this up, I'm pretty pleased with the result. Git config should now be much simpler to understand.

**Review by msfjarvis — DISMISSED** [review]
Everything looks correct

**Comment by fmeum** [comment]
> Everything looks correct

Everything except the `MigrationTest`, that is.

**Comment by msfjarvis** [comment]
> > Everything looks correct
> 
> Everything except the `MigrationTest`, that is.

heh

**Review by msfjarvis — DISMISSED** [review]
Hopefully this is the end of my embarassments

File: app/src/main/java/com/zeapo/pwdstore/Migrations.kt [hunk 4]
```diff
@@ -30,24 +30,13 @@ private fun migrateToGitUrlBasedConfig(context: Context) {
...
-                // Only absolute paths are supported with custom ports.
-                if (!serverPath.startsWith('/'))
-                    null
-                else
-                    // We have to specify the ssh scheme as this is the only way to pass a custom
-                    // port.
-                    "ssh://$userPart$hostnamePart$portPart$serverPath"
-            }
+            val serverPart = if (serverPath.startsWith('/')) serverPath else "/$serverPath"
+            "ssh://$userPart$hostnamePart:$portPart$serverPart"
```
Comment (fmeum) [inline app/src/main/java/com/zeapo/pwdstore/Migrations.kt]: The different ssh URLs confuse me to no end, but doesn't this change reintroduce the issue with relative paths for the standard port? Is there a way we can leave the migration code as is and just fix up the test instead?

Comment (msfjarvis) [reply app/src/main/java/com/zeapo/pwdstore/Migrations.kt]: Ugh I knew I was missing something...

**Comment by msfjarvis** [comment]
Reverted my latest commit, did some digging and the failure is being caused by `GitSettings#updateConnectionSettingsIfValid` returning `MissingUsername`.

**Comment by msfjarvis** [comment]
```diff
--- app/src/androidTest/java/com/zeapo/pwdstore/MigrationsTest.kt
+++ app/src/androidTest/java/com/zeapo/pwdstore/MigrationsTest.kt
@@ -62,7 +62,7 @@ class MigrationsTest {
         checkOldKeysAreRemoved(context)
         assertEquals(
             context.sharedPrefs.getString(PreferenceKeys.GIT_REMOTE_URL),
-            "msfjarvis@192.168.0.102:/mnt/disk3/pass-repo"
+            "ssh://msfjarvis@192.168.0.102/mnt/disk3/pass-repo"
         )
     }
 
--- app/src/main/java/com/zeapo/pwdstore/Migrations.kt
+++ app/src/main/java/com/zeapo/pwdstore/Migrations.kt
@@ -38,7 +38,7 @@ private fun migrateToGitUrlBasedConfig(context: Context) {
             val portPart =
                 if (serverPort == "22" || serverPort.isEmpty()) "" else ":$serverPort"
             if (portPart.isEmpty()) {
-                "$userPart$hostnamePart:$serverPath"
+                "ssh://$userPart$hostnamePart$serverPath"
             } else {
                 // Only absolute paths are supported with custom ports.
                 if (!serverPath.startsWith('/'))
```

This fixes the tests but breaks the relative path scenario.

**Comment by fmeum** [comment]
> ```diff
> --- app/src/androidTest/java/com/zeapo/pwdstore/MigrationsTest.kt
> +++ app/src/androidTest/java/com/zeapo/pwdstore/MigrationsTest.kt
> @@ -62,7 +62,7 @@ class MigrationsTest {
>          checkOldKeysAreRemoved(context)
>          assertEquals(
>              context.sharedPrefs.getString(PreferenceKeys.GIT_REMOTE_URL),
> -            "msfjarvis@192.168.0.102:/mnt/disk3/pass-repo"
> +            "ssh://msfjarvis@192.168.0.102/mnt/disk3/pass-repo"
>          )
>      }
>  
> --- app/src/main/java/com/zeapo/pwdstore/Migrations.kt
> +++ app/src/main/java/com/zeapo/pwdstore/Migrations.kt
> @@ -38,7 +38,7 @@ private fun migrateToGitUrlBasedConfig(context: Context) {
>              val portPart =
>                  if (serverPort == "22" || serverPort.isEmpty()) "" else ":$serverPort"
>              if (portPart.isEmpty()) {
> -                "$userPart$hostnamePart:$serverPath"
> +                "ssh://$userPart$hostnamePart$serverPath"
>              } else {
>                  // Only absolute paths are supported with custom ports.
>                  if (!serverPath.startsWith('/'))
> ```
> 
> This fixes the tests but breaks the relative path scenario.

Let me check why URIish is broken here.

File: app/src/main/java/com/zeapo/pwdstore/git/config/GitSettings.kt [hunk 5]
```diff
@@ -104,23 +97,39 @@ object GitSettings {
...
-    fun updateConnectionSettingsIfValid(newProtocol: Protocol, newAuthMode: AuthMode, newUrl: String, newBranch: String): UpdateConnectionSettingsResult {
+    fun updateConnectionSettingsIfValid(newAuthMode: AuthMode, newUrl: String, newBranch: String): UpdateConnectionSettingsResult {
         val parsedUrl = try {
             URIish(newUrl)
         } catch (_: Exception) {
             return UpdateConnectionSettingsResult.FailedToParseUrl
         }
+        val newProtocol = when (parsedUrl.scheme) {
+            in listOf("http", "https") -> Protocol.Https
+            in listOf("ssh") -> Protocol.Ssh
```
Comment (fmeum) [inline app/src/main/java/com/zeapo/pwdstore/git/config/GitSettings.kt]:
```suggestion
            in listOf("ssh", null) -> Protocol.Ssh
```

**Comment by fmeum** [comment]
This breakage seems to be caused by `updateConnectionSettingsIfValid` returning `FailedToParseUrl` since the scheme is `null`. The `listOf("ssh")` in the `when` statement should be extended to include `null`, which I just added.

LGTM if the tests pass.

**Review by msfjarvis — DISMISSED** [review]

**Comment by msfjarvis** [comment]
Confirmed tests pass locally

**Review by msfjarvis — APPROVED** [review]

**Pull request merged**